descriptor wallet: Cache last hardened xpub and use in normalized descriptors

achow101 opened the pull request:
Currently fetching a normalized descriptor requires the wallet to be unlocked as it needs the private keys to derive the last hardened xpub. This is not very user friendly as normalized descriptors shouldn't require and don't involve the private keys except for derivation. We solve this problem by caching the last hardened xpub (which has to be derived at some point when generating the address pool).

However the last hardened xpub was not already being cached. We only cached the immediate parent xpub and derived child keys. For example, with a descriptor derivation path of `/84'/0'/0'/0/*`, the parent xpub that is cached is `m/84'/0'/0'/0`, and the child keys of `m/84'/0'/0'/0/i` (note that child keys would not be cached in this case). This parent xpub is not suitable for the normalized descriptor form as we want the key at `m/84'/0'/0'`. So this PR adds another field to `DescriptorCache` to cache the last hardened xpub so that we can use them for normalized descriptors.

Since `DescriptorCache` is changing, existing descriptor wallets need to be upgraded to use this new cache. The upgrade will occur in the background either at loading time (if the wallet is not encrypted) or at unlocking time in the same manner that `UpgradeKeyMetadata` operates. It will use a new wallet flag `WALLET_FLAG_LAST_HARDENED_XPUB_CACHED` to indicate whether the descriptor wallet has the last hardened xpub cache.

Lastly `listdescriptors` will not require the wallet to be locked and `getaddressinfo`'s `parent_desc` will always be output (assuming the upgrade has occurred).

achow101 committed:
Pass in DescriptorCache to ToNormalizedString

Use the descriptor xpub cache in ToNormalizedString so that the wallet
does not need to be unlocked in order to get the normalized descriptor.

achow101 committed:
descriptors: Cache last hardened xpub

Cache the last hardenex xpub in the DescriptorCache

achow101 committed:
wallet: Store last hardened xpub cache

achow101 committed:
wallet: Upgrade existing descriptor caches

Add functions to upgrade existing descriptor caches to support the use
of last hardened xpub caching.

achow101 committed:
wallet, rpc: listdescriptors does not need unlocked

With the last hardened xpub cache, we don't neeed to have the wallet be
unlocked for listdescriptors.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -410,7 +411,7 @@ class BIP32PubkeyProvider final : public PubkeyProvider
         return true;
     }
-    bool ToNormalizedString(const SigningProvider& arg, std::string& out, bool priv) const override
+    bool ToNormalizedString(const SigningProvider& arg, std::string& out, bool priv, std::optional<DescriptorCache> cache) const override
```
Passing a copy of the entire cache seems overkill here, and especially with std::move's you're going to be making copies in every step in the call graph.

The extreme version would be `std::optional<std::reference_wrapper<const DescriptorCache>>` to achieve something like an optional reference to a const cache. But just `const DescriptorCache*` works just as well...

sipa replied:
Argh, I meant to click comment, not approve...

achow101 replied:
Done

S3RK commented:
Started reviewing this

S3RK commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2291,3 +2291,43 @@ bool DescriptorScriptPubKeyMan::GetDescriptorString(std::string& out, bool priv)
+    FlatSigningProvider out_keys;
+    std::vector<CScript> scripts_temp;
+    DescriptorCache temp_cache;
+    if (!m_wallet_descriptor.descriptor->Expand(0, provider, scripts_temp, out_keys, &temp_cache)){
```
Why do we need a temp cache?

achow101 replied:
To avoid modifying the wallet state if something goes wrong during expansion. It is possible that the cache could be modified, but something else causes expansion to fail.

S3RK replied:
I don't understand the real life scenario. From what I see `Expand` fails if we can't get a pub key. For example we tried to derive hardened derivation from the cache or the cache itself was inconsistent. This seems to indicate some logic error in the code itself or a corrupted state. Is my understanding correct or do I miss something?

achow101 replied:
Consider `multi(2,xpub.../0/*,xpub.../0h/*)`. The first xpub is entirely unhardened derivation. So we will have entries added to the cache when it is expanded. When `ExpandHelper` moves onto the second xpub, it finds it is entirely hardened derivation. If the keys are unavailable (e.g. they're encrypted), expand now fails here. However the cache now has entries for the first xpub but no entries for the second. While this is not a dangerous state to be in, the in memory state will not be accurately reflected on disk and that is generally something that I think we should avoid. Thus the temp cache is used.

S3RK reviewed (commented):
Wow, it took more than expected to grasp what's going on. I think I still need more time, but I have a few questions/suggestions now already.

achow101 commented on src/script/descriptor.h:
```diff
@@ -110,7 +110,7 @@ struct Descriptor {
 
     /** Convert the descriptor to a normalized string. Normalized descriptors have the xpub at the last hardened step. This fails if the provided provider does not have the private keys to derive that xpub. */
-    virtual bool ToNormalizedString(const SigningProvider& provider, std::string& out, bool priv) const = 0;
+    virtual bool ToNormalizedString(const SigningProvider& provider, std::string& out, bool priv, const DescriptorCache* cache = nullptr) const = 0;
```
Not every caller of this will necessarily have a cache to pass.

S3RK replied:
Ok. I don't have strong feelings about it, but right now the only caller besides the unit tests is `DescriptorScriptPubKeyMan`. Though I can imagine other possibilities in the future I'd still go with more restrictive signature as it's easier to relax the requirements rather than the other way around.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2291,3 +2291,43 @@ bool DescriptorScriptPubKeyMan::GetDescriptorString(std::string& out, bool priv)
+    uint256 id = GetID();
+    for (const auto& lh_xpub_pair : temp_cache.GetCachedLastHardenedExtPubKeys()) {
+        CExtPubKey xpub;
+        if (m_wallet_descriptor.cache.GetCachedLastHardenedExtPubKey(lh_xpub_pair.first, xpub)) {
```
I don't follow.

This is checking to see if elements in the temp cache are in the stored cache;

S3RK replied:
But the stored cache is always empty, no? I'm looking at the line 2303

achow101 replied:
Ah, I see. There shouldn't be anything in the stored cache, however I am doing this as a belt-and-suspenders check.

S3RK replied:
I find it confusing when we check same conditions multiple times in different places because it's not clear what is the responsibility/assumptions of each class/function. It also makes it harder to modify the code. `DescriptorScriptPubKeyMan` and `CWallet` are strongly coupled anyway.

Maybe we can just remove the check on the top of the function? Even in the worst case nothing will go wrong 1) If wallet is locked then expand will fail 2) if the cache exists already we will just regenerate it.

achow101 replied:
This is largely just code duplicated from TopUp. The intention was to refactor them to avoid the duplication, but I never got around to doing that.

The check at the top of the function is important to avoid unnecessary extra computation which slows down wallet loading and unlocking.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -432,29 +432,39 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        // Get the path to the last hardened stup
         KeyOriginInfo origin;
         int k = 0;
         for (; k <= i; ++k) {
```
I don't think that's really necessary since we only do the split in this one spot.

S3RK replied:
Feel free to leave it as it. But just to clarify my point. It's not about code deduplication, it's about readability and comprehension. There are 3 loops dedicated to splitting derivation path and I think they distract from the main purpose of this function.

S3RK reviewed (commented):
Code review ACK 1cb501721ecbd2e3a86f16f4ed9db4cf8ac6ba3f
I don't see any logical errors in the code, it does what it supposed to do.

With that said, I don't like the direction in which `DescriptorImpl::ToStringHelper` and `DescriptorScriptPubKeyMan::TopUp` is going. The code becomes more and more complicated and there are untapped opportunities for simplification.

I believe refactoring should go together with feature implementation because of the following reasons:
* just a pure refactoring PR by itself will get much less attention and therefore lower chances to get merged
* this is not an urgent feature and it can wait for a refactoring to be implemented 

I'm happy to propose, discuss and implement some refactoring ideas if you think it will benefit the project.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1817,6 +1817,19 @@ bool DescriptorScriptPubKeyMan::TopUp(unsigned int size)
+                }
+                continue;
+            }
+            if (!batch.WriteDescriptorLastHardenedCache(lh_xpub_pair.second, id, lh_xpub_pair.first)) {
```
`LoadWallet`, which is called as part of wallet creation, will always call `CWallet::UpgradeDescriptorCache`. Since a newly created wallet is never encrypted at that stage, and has no `ScriptPubKeyMan`s, `UpgradeDescriptorCache` will just set the flag.

S3RK commented:
reACK ec22064. Just rebased on top of latest master

My [comment above](https://github.com/bitcoin/bitcoin/pull/21329#pullrequestreview-614021474) is still relevant though.

achow101 commented:
> With that said, I don't like the direction in which `DescriptorImpl::ToStringHelper` and `DescriptorScriptPubKeyMan::TopUp` is going. The code becomes more and more complicated and there are untapped opportunities for simplification.

What direction do you think it should be going in?

S3RK commented:
> What direction do you think it should be going in?

Thanks for asking. I think we need to try to simplify the code for readability and comprehension. Some rough ideas:

1) for `ToStringHelper` we can a) avoid multiple flags controlling the control flow b) shift non-general code to child classes (addr, raw, multisig). 

To be more specific for this PR: I think it's better to have `ToNormalizedString` reimplement some logic and not call `ToStringHelper` at all. Both functions would be easier to reason about. We also don't need to support normalized private descriptors case, no code actually uses this.

2) for `DescriptorScriptPubKeyMan::TopUp` we can keep the code on the same/constant level of abstraction  by better encapsulating cache structure. One option is to introduce something like `DescriptorCache::CopyFrom(DescriptorCache)` and `WalletBatch::WriteCache(DescriptorCache)`.

achow101 committed:
Refactor Cache merging and writing

Instead of having a large blob of cache merging code in TopUp, refactor
this into DescriptorCache so that it can merge and provide a diff
(another DescriptorCache containing just the items that were added).
Then TopUp can just write everything that was in the diff.

achow101 committed:
Move DescriptorCache writing to WalletBatch

Instead of adhoc writing of the items in DescriptorCache, move it all
into WalletBatch.

achow101 committed:
Remove priv option for ToNormalizedString

achow101 committed:
Change DescriptorImpl::ToStringHelper to use an enum

Instead of having multiple, possibly conflicting, bools controlling the
flow of ToStringHelper, use an enum.

achow101 commented:
> To be more specific for this PR: I think it's better to have `ToNormalizedString` reimplement some logic and not call `ToStringHelper` at all. Both functions would be easier to reason about. We also don't need to support normalized private descriptors case, no code actually uses this.

I kind of disagree. All of the `To*String` functions do roughly the same thing. The construction of all of the non-key related parts are all the same and it doesn't make sense to me to be duplicating that. The only parts that are different are the key things, which is why `ToStringHelper` has options in order to determine which key function to use. I think that instead of using bools, it would be better to use an enum, so I have implemented it that way.

> for `DescriptorScriptPubKeyMan::TopUp` we can keep the code on the same/constant level of abstraction  by better encapsulating cache structure. One option is to introduce something like `DescriptorCache::CopyFrom(DescriptorCache)` and `WalletBatch::WriteCache(DescriptorCache)`.

I've added ` DescriptorCache::MergeAndDiff` which merges another `DescriptorCache` with the current `DescriptorCache` and returns another `DescriptorCache` containing a copy of the new items. I've also added `WalletBatch::WriteDescriptorCache` which takes a `DescriptorCache` and writes all of the items. This cleans up `TopUp` and deduplicates some code.

S3RK commented:
utACK e668b9e. Thanks for incorporating my suggestions.

jonatack commented:
Reviewing.

jonatack commented on src/script/descriptor.cpp:
```diff
@@ -287,9 +286,6 @@ class BIP32PubkeyProvider final : public PubkeyProvider
     DeriveType m_derive;
-    // Cache of the parent of the final derived pubkeys.
-    // Primarily useful for situations when no read_cache is provided
-    CExtPubKey m_cached_xpub;
```
0cc90f1e3dbccd5940b05b633260fdcd8dac5581 essentially reverts 09e25071f40 IIUC. Perhaps re-add `const` to the four `GetPubKey()` member functions in the commit and mention in the commit message that it reverts 09e25071f40

```diff
-    virtual bool GetPubKey(int pos, const SigningProvider& arg, CPubKey& key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) = 0;
+    virtual bool GetPubKey(int pos, const SigningProvider& arg, CPubKey& key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const = 0;
```

jonatack replied:
Apart from this, ACK up to 8fac319f6d180fa5c597ea08beb7827331e7d9ec so far (the first three commits).

achow101 replied:
I've made this commit into a pure revert with an extra comment saying why.

jonatack replied:
Cool. Continuing review.

achow101 committed:
Revert "Cache parent xpub inside of BIP32PubkeyProvider"

This reverts commit 09e25071f40c564af08a1386c39c4f2d8eb484b6.

The changes made in this commit have turned out to be unnecessary and
confusing, so it is being reverted.

jonatack reviewed (commented):
Semi ACK e6cf0ed92de31a5ac35a271b0da8f0a8364d1175 reviewed, debug-built and ran unit tests and some of the descriptor functional tests at each commit. I'm not very familiar with this code and it could be clearer to the uninitiated IMHO, so I'm not confident enough to give a full ACK. Various minor suggestions follow, most of them for readability, feel free to pick and choose.

fjahr commented:
tACK e6cf0ed92de31a5ac35a271b0da8f0a8364d1175

Reviewed code, light manual testing with restarts to ensure persistence works (could have been covered by a test case as well).

nit: in d87b544b834077f102724415e0fada6ee8b2def2 there is a typo in commit description: "hardenex"

achow101 commented:
I will implement the suggestions if I have to retouch.

S3RK commented:
reACK e6cf0ed

Changes from last review: 1) `PubkeyProvider::GetPubKey` became const 2) necessary changes for taproot descriptors

Happy to reACK again if you want to incorporate jonatack's readability suggestions

fanquake commented:
@meshcollider

meshcollider reviewed (commented):
Code review + functional test run ACK e6cf0ed92de31a5ac35a271b0da8f0a8364d1175

meshcollider commented:
Nits can be left for follow-up if needed, lets just get this in

achow101 commented:
This was merged, but github isn't detecting it.

jonatack commented:
There have been several of these, I think this is the third one in as many days.
